### PR TITLE
crop scrollbar in screenshots(closes #3292)

### LINKED ISSUE #3292: Element screenshots are blank/empty, but element is in viewport and exists
### What is your Test Scenario?
Take a screenshots from multiple elemets from a list of elements inside a testcase.

(I also tryed to seperate each screenshot taken to a single testcase and disabled the 'Use hardware acceleration' of chrome, doesn't work anyway) 

Just fyi: 
In general I use TestCafé in combination with [looksSame](https://github.com/gemini-testing/looks-same)(compare images) to provide screenshot/visual regression - tests with advanced user actions (provided by TestCafé)

### What is the Current behavior?
Nearly half of the screenshots taken with the testcase below are empty/blank (see screenshots)

_It look's like the runner hides the viewport before taking the screenshot, you can reproduce this by running the testcase below in a real browser instead of headless._

### What is the Expected behavior?
All screenshots should display their corresponding element.
 
### What is your web application and your TestCafe test code?

 
This page provides a matching example case:
www.lottohelden.de/gewinnwahrscheinlichkeiten/lotterien/

<details>
<summary>Test code:</summary>

 
```js
const baseUrl = 'www.lottohelden.de'

fixture(`Run Screenshots for VR-Tests: ${baseUrl}`)
  .page(`${baseUrl}`)
  .beforeEach( async t => {
    await t
      .wait(2500)
      .resizeWindow(1920, 1080)
  })

 test.page(`${baseUrl}/gewinnwahrscheinlichkeiten/lotterien/`)
  ('screenshot desktop tabels', async t => {
  const oddsPanelList = Selector('.current-odds-chance');
  const count = await oddsPanelList.count
  for(let counter = 0; counter < count; counter++) {
    await t
      .wait(3000)
      .takeElementScreenshot(oddsPanelList.nth(counter), `desktop-table${counter}.png`)
  }
})
```
</details>

<details>
<summary>Generated screenshots from the testcase above:</summary>

how it should be:
grafik

where it all starts:
grafik

the issue: 
grafik

</details>

### Steps to Reproduce:

1. Setup testcafé  
3. Execute given testcode ( generate the screenshots )
4. Open image number 9 or higher to see the problem
 
### Your Environment details:
 
* testcafe version:                   0.23.3
* node.js version:                    v10.11.0
* command-line arguments:  testcafe chrome -s "src/screenshots"
* browser name and version: Chrome Version 71
* platform and version:          macOS 10.14

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

File: src/browser/provider/built-in/chrome/cdp.js [hunk 1]
```diff
@@ -45,18 +44,8 @@ async function setEmulation (runtimeInfo) {
...
-        x:      visualViewport.pageX,
-        y:      visualViewport.pageY,
-        width:  visualViewport.clientWidth,
-        height: visualViewport.clientHeight,
-        scale:  visualViewport.scale
-    };
-
-    return await client.Page.captureScreenshot({ fromSurface: true, clip: clipRegion });
+export async function getScreenshotData (client) {
+    return await client.Page.captureScreenshot({ format: 'png', fromSurface: true });
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/cdp.js]: According to https://chromedevtools.github.io/devtools-protocol/tot/Page, `png` is the default value for the `format` parameter and `true` is the default value for `fromSurface` parameter.
It means you can omit them.

File: src/browser/provider/built-in/chrome/index.js [hunk 2]
```diff
@@ -64,9 +64,19 @@ export default {
     },
 
     async takeScreenshot (browserId, path) {
-        const runtimeInfo = this.openedBrowsers[browserId];
-
-        await cdp.takeScreenshot(path, runtimeInfo);
+        const runtimeInfo    = this.openedBrowsers[browserId];
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/index.js]: const { client } = runtimeInfo;

File: src/screenshots/crop.js [hunk 3]
```diff
@@ -50,7 +47,7 @@ function markSeedToId (markSeed) {
     return id;
 }
 
-function detectClippingArea (srcImage, { markSeed, clientAreaDimensions, cropDimensions, screenshotPath } = {}) {
+function detectClippingArea (srcImage, markSeed, clientAreaDimensions, cropDimensions) {
```
Comment (miherlosev) [inline src/screenshots/crop.js]: Why do you ungroup the parameters?

File: src/screenshots/crop.js [hunk 4]
```diff
@@ -112,22 +109,33 @@ function copyImagePart (srcImage, { left, top, width, height }) {
...
-    const dstImage = copyImagePart(srcImage, clippingArea);
+export async function cropScreenshotByPath (path, markSeed, clientAreaDimensions, cropDimensions) {
+    const sourceImage = await readFile(path);
 
-    await writePng(screenshotPath, dstImage);
+    try {
+        await cropScreenshotBinary(path, markSeed, clientAreaDimensions, cropDimensions, sourceImage);
+    }
+    catch (err) {
+        deleteFile(path);
```
Comment (miherlosev) [inline src/screenshots/crop.js]: await deleteFile(path)

File: src/screenshots/crop.js (hunk 3 above)
Comment (AlexKamaev) [reply src/screenshots/crop.js]: I found it unreasonable to construct an object to pass it as an argument, and then immediately destructure this object. Using parameters in the usual manner are sufficient here.

**Comment by AlexKamaev** [comment]
FPR

File: src/browser/provider/built-in/chrome/index.js [hunk 5]
```diff
@@ -64,9 +64,18 @@ export default {
...
+        const screenshotData = await cdp.getScreenshotData(client);
+        const binaryImage    = Buffer.from(screenshotData.data, 'base64');
+
+        const { visualViewport } = await client.Page.getLayoutMetrics();
+
+        await cropScreenshotBinary(path, false, null, {
+            right:  visualViewport.clientWidth,
+            left:   0,
+            top:    0,
+            bottom: Number.MAX_VALUE
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/index.js]: What about horizontal scrollbars?

File: src/browser/provider/built-in/chrome/cdp.js [hunk 6]
```diff
@@ -45,18 +44,8 @@ async function setEmulation (runtimeInfo) {
...
-        x:      visualViewport.pageX,
-        y:      visualViewport.pageY,
-        width:  visualViewport.clientWidth,
-        height: visualViewport.clientHeight,
-        scale:  visualViewport.scale
-    };
-
-    return await client.Page.captureScreenshot({ fromSurface: true, clip: clipRegion });
+export async function getScreenshotData (client) {
+    return await client.Page.captureScreenshot({ fromSurface: true });
```
[the remark this answers is not in the document]
Comment (AlexKamaev) [reply src/browser/provider/built-in/chrome/cdp.js]: ok, it's default value

File: src/browser/provider/built-in/chrome/index.js (hunk 5 above)
Comment (AlexKamaev) [reply src/browser/provider/built-in/chrome/index.js]: thanks, I completely forgot about them

File: src/browser/provider/built-in/chrome/index.js [hunk 7]
```diff
@@ -64,9 +64,19 @@ export default {
...
 
     async takeScreenshot (browserId, path) {
-        const runtimeInfo = this.openedBrowsers[browserId];
-
-        await cdp.takeScreenshot(path, runtimeInfo);
+        const { client }     = this.openedBrowsers[browserId];
+        const screenshotData = await cdp.getScreenshotData(client);
+        const binaryImage    = Buffer.from(screenshotData.data, 'base64');
+
+        const { visualViewport }            = await client.Page.getLayoutMetrics();
```
Comment (AndreyBelym) [inline src/browser/provider/built-in/chrome/index.js]: Please keep all CDP calls in the `cdp` module.

Comment (AndreyBelym) [reply src/browser/provider/built-in/chrome/index.js]: BTW we need this cropping logic for videos too, so maybe we should move it to the `cdp.getScreenshotData`?

File: src/screenshots/crop.js [hunk 8]
```diff
@@ -112,22 +109,33 @@ function copyImagePart (srcImage, { left, top, width, height }) {
...
+    if (clippingArea.width <= 0 || clippingArea.height <= 0)
         throw new InvalidElementScreenshotDimensionsError(clippingArea.width, clippingArea.height);
-    }
 
-    if (!markSeed && !cropDimensions)
-        return true;
+    if (markSeed || cropDimensions) {
+        png = copyImagePart(png, clippingArea);
+
+        await writePng(path, png);
```
Comment (AndreyBelym) [inline src/screenshots/crop.js]: I think it's better to extract all file-related logic to a different place and keep `cropScreenshotBinary` for image operations only.

File: src/screenshots/crop.js [hunk 9]
```diff
@@ -64,7 +61,7 @@ function detectClippingArea (srcImage, { markSeed, clientAreaDimensions, cropDim
         const markIndex = srcImage.data.indexOf(mark);
 
         if (markIndex < 0)
-            throw new Error(renderTemplate(WARNING_MESSAGES.screenshotMarkNotFound, screenshotPath, markSeedToId(markSeed)));
+            return null;
```
Comment (AndreyBelym) [inline src/screenshots/crop.js]: We should emit the warning if we can't find a mark on screenshots that contain browser UI. I guess we need to extract `crop by coordinates` code and make it usable by the `chome` provider without for searching for a mark.

**Comment by AlexKamaev** [comment]
FPR

File: src/screenshots/png.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
+import fs from 'fs';
```
Comment (miherlosev) [inline src/screenshots/png.js]: This file should be called as `utils.js`.

File: src/screenshots/crop.js [hunk 11]
```diff
@@ -50,84 +15,83 @@ function markSeedToId (markSeed) {
...
+        clipTop:    0
+    };
 
-    for (let i = 0; i < height; i++) {
-        const srcStartIndex = (srcImage.width * (i + top) + left) * MARK_BYTES_PER_PIXEL;
+    let markLineNumber = null;
 
-        srcImage.data.copy(dstImage.data, stride * i, srcStartIndex, srcStartIndex + stride);
+    if (markSeed && clientAreaDimensions) {
+        clipInfo = updateClipInfoByMarkSeed(pngImage, path, markSeed, clientAreaDimensions);
```
Comment (AndreyBelym) [inline src/screenshots/crop.js]: I would like if we refactored it a bit more in the following way:
```js
let markPosition = null;

if (markSeed && clientAreaDimensions) {
    markPosition = calculateMarkPosition(pngImage, markSeed, clientAreaDimensions);

    if (!markPosition)
        throw new Error(...);
 
    clipInfo = updateClipInfoByMarkSeed(pngImage, markPosition, clientAreaDimensions);
}

if (markPosition && markPosition.y === clipInfo.clipBottom)
    clipInfo.clipBottom--;   
```

**Comment by kalabro** [comment]
Thanks so much for the fix @AlexKamaev!

**Pull request merged**